compiler/runtime/OMRRSSReport.cpp

```diff
@@ -0,0 +1,354 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2000
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "runtime/CodeCacheManager.hpp"
+#include "runtime/CodeCache.hpp"
+
+OMR::RSSReport *OMR::RSSReport::_instance = NULL;
+
+const char *OMR::RSSItem::itemNames [] =
+      {
+      "Unknown",
+      "HEADER",
+      "ALIGNMENT",
+      "COLD BLOCKS",
+      "OVERESTIMATE",
+      "OOL",
+      "SNIPPET"
+      };
+
+void
+OMR::RSSRegion::addRSSItem(OMR::RSSItem *item, int32_t threadId, const char* methodName)
+   {
+#ifdef DEBUG_RSS_REPORT
+   TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding item addr=%p size=%zu type=%s thread=%d method=%s", item->_addr, item->_size,
+                                  OMR::RSSItem::itemNames[item->_type], threadId,
+                                  methodName ? methodName : "no method");
+#endif
+
+   uint8_t *address = item->_addr;
+   TR_PersistentList<TR::DebugCounterAggregation> counters = item->_counters;
+
+   TR_ASSERT_FATAL(address, "Address should not be null");
+   TR_ASSERT_FATAL(_pageSize >=0, "Page size should be set");
+
+   size_t addressPage = (uintptr_t)address / _pageSize;
+   size_t startPage = (uintptr_t)_start / _pageSize;
+   int32_t offset = static_cast<int32_t>((_grows == lowToHigh) ? addressPage - startPage : startPage - addressPage);
+
+   TR_ASSERT_FATAL(offset >= 0, "Offset should be >= 0\n");
+
+   addToListSorted(&_pageMap[offset], item);
+
+   // split across pages if necessary
+   size_t tillPageEnd = _pageSize - ((uintptr_t)address % _pageSize);
```

Comment:
RSSItem can span across multiple pages since it describes an arbitrary range of memory within RSSReport. As in the example I gave below: cold cache region can contain many items: cold blocks, snippets, OOL. Each of then can be bigger than a page. On the other hand, the system reports RSS status per page. That's why we have an array pf pages and each page has a list of items.

Reply:
Is this a correct understanding:

`RSSItem` can span across multiple pages when it is instantiated, but once you call `addRSSItem`, the `RSSItem` is only going to span a page.

Reply:
Correct, essentially we will split the item provided by the consumer into multiple items. So the original one will be modified.